File: 366-Find-Leaves-of-Binary-Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var findLeaves = function(root) {
    var ans = [];
    
    var helper = function(root){
        if(root === null)
            return 0;
        var left = helper(root.left);
        var right = helper(root.right);
        var level = Math.max(left, right);
        if(ans[level] === undefined)
            ans[level] = [];
        ans[level].push(root.val);
        return level+1;
    }
    helper(root);
    return ans;
};

File: 366-Find-Leaves-of-Binary-Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    List<List<Integer>> ans;
    public List<List<Integer>> findLeaves(TreeNode root) {
        //a tree node's level is the max height count from its leafs
        
        ans = new ArrayList<>();
        helper(root);
        return ans;
    }
    
    public int helper(TreeNode root){
        if(root == null)    return 0;
        int left = helper(root.left);
        int right = helper(root.right);
        int level = Math.max(left, right);
        if(ans.size() <= level){
            for(int i = ans.size(); i <= level; i++)
                ans.add(new ArrayList<Integer>());
        }
        ans.get(level).add(root.val);
        return level+1;
    }
}